patches/minecraft/net/minecraft/tags/TagLoader.java.patch

```diff
@@ -1,11 +1,108 @@
+-      ImmutableSet.Builder<T> builder = ImmutableSet.builder();
++      return buildSets(p_215979_, p_215980_).mapRight(p -> p.additive());
++   }
++
++   // FORGE: Build both the additive and subtractive sets for the given tags
++   private Either<Collection<TagLoader.EntryWithSource>, AdditiveSubtractiveSets<T>> buildSets(TagEntry.Lookup<T> lookup, List<TagLoader.EntryWithSource> entries) {
++      ImmutableSet.Builder<T> additive = ImmutableSet.builder();
++      ImmutableSet.Builder<T> subtractive = ImmutableSet.builder();
+       List<TagLoader.EntryWithSource> list = new ArrayList<>();
+ 
+-      for(TagLoader.EntryWithSource tagloader$entrywithsource : p_215980_) {
+-         if (!tagloader$entrywithsource.f_216042_().m_215927_(p_215979_, builder::add)) {
++      for(TagLoader.EntryWithSource tagloader$entrywithsource : entries) {
++         if (!tagloader$entrywithsource.f_216042_().m_215927_(lookup, tagloader$entrywithsource.additive() ? additive::add : subtractive::add)) {
+             list.add(tagloader$entrywithsource);
+          }
+       }
+ 
+-      return list.isEmpty() ? Either.right(builder.build()) : Either.left(list);
++      return list.isEmpty() ? Either.right(new AdditiveSubtractiveSets<>(additive.build(), subtractive.build())) : Either.left(list);
+    }
+ 
+    public Map<ResourceLocation, Collection<T>> m_203898_(Map<ResourceLocation, List<TagLoader.EntryWithSource>> p_203899_) {
+-      final Map<ResourceLocation, Collection<T>> map = Maps.newHashMap();
++      // FORGE: Store the full information provided from the private build method
++      final Map<ResourceLocation, AdditiveSubtractiveSets<T>> map = Maps.newHashMap();
+       TagEntry.Lookup<T> lookup = new TagEntry.Lookup<T>() {
+          @Nullable
+          public T m_213619_(ResourceLocation p_216039_) {
+@@ -142,7 +_,9 @@
+ 
+          @Nullable
+          public Collection<T> m_214048_(ResourceLocation p_216041_) {
+-            return map.get(p_216041_);
++            // FORGE: Only return the additive sets
++            if (map.get(p_216041_) == null) { return null; }
++            return map.get(p_216041_).additive();
+          }
+       };
+       Multimap<ResourceLocation, ResourceLocation> multimap = HashMultimap.create();
+@@ -163,23 +_,40 @@
+       Set<ResourceLocation> set = Sets.newHashSet();
        p_203899_.keySet().forEach((p_216008_) -> {
           m_144523_(p_203899_, multimap, set, p_216008_, (p_215984_, p_215985_) -> {
-             this.m_215978_(lookup, p_215985_).ifLeft((p_215977_) -> {
+-            this.m_215978_(lookup, p_215985_).ifLeft((p_215977_) -> {
 -               f_13445_.error("Couldn't load tag {} as it is missing following references: {}", p_215984_, p_215977_.stream().map(Objects::toString).collect(Collectors.joining(", ")));
++            this.buildSets(lookup, p_215985_).ifLeft((p_215977_) -> {
++               // FORGE: Log the references on separate lines for easier visibility
```

Comment:
Unnecessary comment.

Reply:
I was asked to add this by @sciwhiz12

Reply:
Can confirm, this one should be left in.